src/gui/sdlmain.cpp

```diff
@@ -356,64 +359,73 @@ static double get_host_refresh_rate()
 			            sdl_rate);
 			return RefreshRateHostDefault;
 		}
+
 		assert(sdl_rate >= RefreshRateMin);
 		return sdl_rate;
 	};
 
 	auto get_vrr_rate = [](const int sdl_rate) {
-		constexpr auto vrr_backoff_hz = 3;
-		return sdl_rate - vrr_backoff_hz;
+		constexpr auto VrrBackoffHz = 3;
+		return sdl_rate - VrrBackoffHz;
 	};
 
 	auto get_sdi_rate = [](const int sdl_rate) {
-		const auto is_odd = sdl_rate % 2 != 0;
-		const auto not_div_by_5 = sdl_rate % 5 != 0;
+		// TODO needs explanation - what are we even calculating here?
+		const auto is_odd           = sdl_rate % 2 != 0;
+		const auto not_div_by_5     = sdl_rate % 5 != 0;
 		const auto next_is_div_by_3 = (sdl_rate + 1) % 3 == 0;
+
 		const bool should_adjust = is_odd && not_div_by_5 && next_is_div_by_3;
 		constexpr auto sdi_factor = 1.0 - 1.0 / 1000.0;
+
 		return should_adjust ? (sdl_rate + 1) * sdi_factor : sdl_rate;
 	};
```

Comment:
SDI rates are (usually) fractional: 23.98, 24, 25, 29.97, 30, 50, 59.94, and 60 frames per second - but this function takes in an integer SDL rate.

Maybe it's doing this conversion from integer to these SDI rates? (and maybe follows some SDI rules that calculate it properly for even more rates in between?).

Will read more later about it.

Reply:
That's a possibility, but in that case I'd just use a simple table approach instead of this.

Reply:
Did a quick dump from Python.

My guess is it's generalized to handle all the potential rates that monitors can and will (eventually) throw at it.  It would be even better if SDL could give the display rate as a float or double :face_with_head_bandage: (fingers crossed SDL3 does!)

```python
def get_sdi_rate(sdl_rate):
    is_odd = sdl_rate % 2 != 0
    not_div_by_5 = sdl_rate % 5 != 0
    next_is_div_by_3 = (sdl_rate + 1) % 3 == 0
    
    should_adjust = is_odd and not_div_by_5 and next_is_div_by_3
    sdi_factor = 1.0 - 1.0 / 1000.0
    
    if should_adjust:
        return (sdl_rate + 1) * sdi_factor
    else:
        return sdl_rate

for i in range(10, 301):
    print(i, get_sdi_rate(i))
```

| SDL input (Hz) | SDI rate (Hz) |
|:--- |:--- | 
| 10 | 10 |
| 11 | 11.988 |
| 12 | 12 |
| 13 | 13 |
| 14 | 14 |
| 15 | 15 |
| 16 | 16 |
| 17 | 17.982 |
| 18 | 18 |
| 19 | 19 |
| 20 | 20 |
| 21 | 21 |
| 22 | 22 |
| 23 | 23.976 |
| 24 | 24 |
| 25 | 25 |
| 26 | 26 |
| 27 | 27 |
| 28 | 28 |
| 29 | 29.97 |
| 30 | 30 |
| 31 | 31 |
| 32 | 32 |
| 33 | 33 |
| 34 | 34 |
| 35 | 35 |
| 36 | 36 |
| 37 | 37 |
| 38 | 38 |
| 39 | 39 |
| 40 | 40 |
| 41 | 41.958 |
| 42 | 42 |
| 43 | 43 |
| 44 | 44 |
| 45 | 45 |
| 46 | 46 |
| 47 | 47.952 |
| 48 | 48 |
| 49 | 49 |
| 50 | 50 |
| 51 | 51 |
| 52 | 52 |
| 53 | 53.946 |
| 54 | 54 |
| 55 | 55 |
| 56 | 56 |
| 57 | 57 |
| 58 | 58 |
| 59 | 59.94 |
| 60 | 60 |
| 61 | 61 |
| 62 | 62 |
| 63 | 63 |
| 64 | 64 |
| 65 | 65 |
| 66 | 66 |
| 67 | 67 |
| 68 | 68 |
| 69 | 69 |
| 70 | 70 |
| 71 | 71.928 |
| 72 | 72 |
| 73 | 73 |
| 74 | 74 |
| 75 | 75 |
| 76 | 76 |
| 77 | 77.922 |
| 78 | 78 |
| 79 | 79 |
| 80 | 80 |
| 81 | 81 |
| 82 | 82 |
| 83 | 83.916 |
| 84 | 84 |
| 85 | 85 |
| 86 | 86 |
| 87 | 87 |
| 88 | 88 |
| 89 | 89.91 |
| 90 | 90 |
| 91 | 91 |
| 92 | 92 |
| 93 | 93 |
| 94 | 94 |
| 95 | 95 |
| 96 | 96 |
| 97 | 97 |
| 98 | 98 |
| 99 | 99 |
| 100 | 100 |
| 101 | 101.898 |
| 102 | 102 |
| 103 | 103 |
| 104 | 104 |
| 105 | 105 |
| 106 | 106 |
| 107 | 107.892 |
| 108 | 108 |
| 109 | 109 |
| 110 | 110 |
| 111 | 111 |
| 112 | 112 |
| 113 | 113.886 |
| 114 | 114 |
| 115 | 115 |
| 116 | 116 |
| 117 | 117 |
| 118 | 118 |
| 119 | 119.88 |
| 120 | 120 |
| 121 | 121 |
| 122 | 122 |
| 123 | 123 |
| 124 | 124 |
| 125 | 125 |
| 126 | 126 |
| 127 | 127 |
| 128 | 128 |
| 129 | 129 |
| 130 | 130 |
| 131 | 131.868 |
| 132 | 132 |
| 133 | 133 |
| 134 | 134 |
| 135 | 135 |
| 136 | 136 |
| 137 | 137.862 |
| 138 | 138 |
| 139 | 139 |
| 140 | 140 |
| 141 | 141 |
| 142 | 142 |
| 143 | 143.856 |
| 144 | 144 |
| 145 | 145 |
| 146 | 146 |
| 147 | 147 |
| 148 | 148 |
| 149 | 149.85 |
| 150 | 150 |
| 151 | 151 |
| 152 | 152 |
| 153 | 153 |
| 154 | 154 |
| 155 | 155 |
| 156 | 156 |
| 157 | 157 |
| 158 | 158 |
| 159 | 159 |
| 160 | 160 |
| 161 | 161.838 |
| 162 | 162 |
| 163 | 163 |
| 164 | 164 |
| 165 | 165 |
| 166 | 166 |
| 167 | 167.832 |
| 168 | 168 |
| 169 | 169 |
| 170 | 170 |
| 171 | 171 |
| 172 | 172 |
| 173 | 173.826 |
| 174 | 174 |
| 175 | 175 |
| 176 | 176 |
| 177 | 177 |
| 178 | 178 |
| 179 | 179.82 |
| 180 | 180 |
| 181 | 181 |
| 182 | 182 |
| 183 | 183 |
| 184 | 184 |
| 185 | 185 |
| 186 | 186 |
| 187 | 187 |
| 188 | 188 |
| 189 | 189 |
| 190 | 190 |
| 191 | 191.808 |
| 192 | 192 |
| 193 | 193 |
| 194 | 194 |
| 195 | 195 |
| 196 | 196 |
| 197 | 197.802 |
| 198 | 198 |
| 199 | 199 |
| 200 | 200 |
| 201 | 201 |
| 202 | 202 |
| 203 | 203.796 |
| 204 | 204 |
| 205 | 205 |
| 206 | 206 |
| 207 | 207 |
| 208 | 208 |
| 209 | 209.79 |
| 210 | 210 |
| 211 | 211 |
| 212 | 212 |
| 213 | 213 |
| 214 | 214 |
| 215 | 215 |
| 216 | 216 |
| 217 | 217 |
| 218 | 218 |
| 219 | 219 |
| 220 | 220 |
| 221 | 221.778 |
| 222 | 222 |
| 223 | 223 |
| 224 | 224 |
| 225 | 225 |
| 226 | 226 |
| 227 | 227.772 |
| 228 | 228 |
| 229 | 229 |
| 230 | 230 |
| 231 | 231 |
| 232 | 232 |
| 233 | 233.766 |
| 234 | 234 |
| 235 | 235 |
| 236 | 236 |
| 237 | 237 |
| 238 | 238 |
| 239 | 239.76 |
| 240 | 240 |
| 241 | 241 |
| 242 | 242 |
| 243 | 243 |
| 244 | 244 |
| 245 | 245 |
| 246 | 246 |
| 247 | 247 |
| 248 | 248 |
| 249 | 249 |
| 250 | 250 |
| 251 | 251.748 |
| 252 | 252 |
| 253 | 253 |
| 254 | 254 |
| 255 | 255 |
| 256 | 256 |
| 257 | 257.742 |
| 258 | 258 |
| 259 | 259 |
| 260 | 260 |
| 261 | 261 |
| 262 | 262 |
| 263 | 263.736 |
| 264 | 264 |
| 265 | 265 |
| 266 | 266 |
| 267 | 267 |
| 268 | 268 |
| 269 | 269.73 |
| 270 | 270 |
| 271 | 271 |
| 272 | 272 |
| 273 | 273 |
| 274 | 274 |
| 275 | 275 |
| 276 | 276 |
| 277 | 277 |
| 278 | 278 |
| 279 | 279 |
| 280 | 280 |
| 281 | 281.718 |
| 282 | 282 |
| 283 | 283 |
| 284 | 284 |
| 285 | 285 |
| 286 | 286 |
| 287 | 287.712 |
| 288 | 288 |
| 289 | 289 |
| 290 | 290 |
| 291 | 291 |
| 292 | 292 |
| 293 | 293.706 |
| 294 | 294 |
| 295 | 295 |
| 296 | 296 |
| 297 | 297 |
| 298 | 298 |
| 299 | 299.7 |
| 300 | 300 |

Reply:
Cool, but I still don't fully get it why do we care about all this SDI stuff. Just to attempt to derive the real fractional rates from the truncated rates reported by SDL? Why is that important? Lots of questions, and a job for later to understand all this 😆

Reply:
> Just to attempt to derive the real fractional rates from the truncated rates reported by SDL? Why is that important?

I think getting the display's actual fractional rate (say, 59.94 Hz instead of SDL's truncated 59 Hz) is important when in the throttled modes (DOS game @ 70 Hz but display @ 59.94 Hz, as an example).

Ok - but what benefit does that that extra fractional chunk give us? 59 vs 59.94 are both a world apart from DOS's desired 70 Hz.

In the throttled mode, the code times the frame presentation down to the decimal value - so there must be some benefit to getting the timing very close to correct (or consequence if we're wrong).

As frames go by, the 59 Hz frames will be later and later with respect to the display's actual 59.94 Hz rate, and eventually they will be so late that the 59.94 rate will "catch up" and the 59 Hz will be early.. and so on until they eventually land "in phase" with each other.

![phase-diff-between-59-and-59 94hz](https://github.com/user-attachments/assets/f8cc9b82-5b56-41ab-8297-320c264e8486)

Probably the consequence of the wrong timing is when vsync is on or off: when it's on, things are halted and when it's off the image is torn. If the timing is correct, then the halting or tear is eliminated?

```python
import numpy as np
import matplotlib.pyplot as plt

def plot_phase_difference(actual_rate, attempted_rate, duration_seconds=10):
    # Calculate the number of actual cycles
    num_actual_cycles = int(duration_seconds * actual_rate)
    
    # Create time array
    t = np.linspace(0, duration_seconds, num_actual_cycles)
    
    # Calculate phase difference
    phase_difference = (t * (attempted_rate - actual_rate)) % 1
    
    # Adjust phase difference to be between -0.5 and 0.5
    phase_difference = np.where(phase_difference > 0.5, phase_difference - 1, phase_difference)
    
    # Convert phase difference to milliseconds
    time_difference_ms = phase_difference * (1000 / actual_rate)
    
    # Plot the results
    plt.figure(figsize=(12, 6))
    plt.plot(t, time_difference_ms)
    plt.title(f'Phase Difference: {actual_rate} Hz (Actual) vs {attempted_rate} Hz (Attempted)')
    plt.xlabel('Time (seconds)')
    plt.ylabel('Time Difference (milliseconds)')
    plt.ylim(-500 / actual_rate, 500 / actual_rate)  # Set y-axis limit to ±half cycle duration
    plt.axhline(y=0, color='r', linestyle='--')  # Add a horizontal line at y=0
    plt.grid(True)
    plt.show()

actual_rate = 59.94
attempted_rate = 59.0
plot_phase_difference(actual_rate, attempted_rate)
```